% \VignetteIndexEntry{An R Package for belief propagation in genotype-phenotype networks}
% \VignetteKeyword{Conditional Gaussian Networks}
% \VignetteKeyword{Probabilistic networks}
% \VignetteKeyword{Graphical Models}
% \VignetteKeyword{Discrete Bayesian Networks}

\documentclass[12pt, a4paper]{article}

\usepackage{Sweave}

% amsmath package, useful for mathematical formulas
\usepackage{amsmath}

%amssymb package, useful for mathematical symbols
\usepackage{amssymb}

%graphicx package, useful for including eps and pdf graphics
% include graphics with the command \includegraphics
\usepackage{graphicx}
\usepackage{epstopdf}

% packages for tables
\usepackage{cite}
\usepackage{algorithm}
\usepackage{lineno}
\usepackage{color}
\usepackage{setspace}
%\double-spacing

%Text layout
\topmargin 0.0cm
\oddsidemargin 0.5cm
\evensidemargin 0.5cm
\textwidth 16cm 
\textheight 21cm


% Bold the 'Figure #' in the caption and separate it with a period
% Captions will be left justified
\usepackage[labelfont=bf,labelsep=period,justification=raggedright]{caption}


\def\pkg#1{\texttt{#1}}
\def\code#1{{\texttt{#1}}}
\def\R{\texttt{R}}


<<echo=FALSE,print=FALSE>>=
require( geneNetBP )
prettyVersion <- packageDescription("geneNetBP")$Version
prettyDate <- format(Sys.Date())
@

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

\author{Janhavi Moharil\\ University at Buffalo}
\title{Belief Propagation in Genotype-Phenotype Networks using the \pkg{geneNetBP} package}
\date{\pkg{geneNetBP} version \Sexpr{prettyVersion} as of \Sexpr{prettyDate}}

\begin{document}
\SweaveOpts{concordance=TRUE}
% \SweaveOpts{concordance=FALSE}

\maketitle

\tableofcontents

\section{Introduction}

\noindent The \code{geneNetBP} package implements methods to predict system-wide changes in beliefs after absorbing evidence in probabilistic graphical models. The package includes functions to fit Conditional Gaussian Bayesian Network (CG-BN) to specifically genotype-phenotype or Quantitative Trait Loci (QTL) data, absorb evidence in these networks and quantify and visualize the changes in network beliefs.\\

\noindent The package makes extensive use of \code{RHugin} package that provides an R interface for the Hugin Decision Engine, a commercial software for building and infering Bayesian belief networks. Note that \code{RHugin} is currently not available on CRAN and is hosted on R-Forge.It can be downloaded from http://rhugin.r-forge.r-project.org. The Hugin Decision Engine can be downloaded from http://www.hugin.com. Detailed installation instructions of the \code{geneNetBP} package and package dependencies are available on the project homepage on R-Forge, http://genenetbp.r-forge.r-project.org.\\ 
 
\section{Datasets}

\noindent There are 3 datasets provided with this package. 

\subsection{mouse} 
\noindent The \emph{Mus Musculus} Kidney eQTL data (\code{mouse}) was obtained from a F2 inner-cross between inbred MRL/MpJ and SM/J strains of mice \cite{hageman2011}. The original data consists of 33,872 gene expression traits for 173 males. After linkage analysis and filtering based on location and significance of QTL, the data consists of 14 genes and their SNP markers corresponding to their QTL. Thus the final dataset has 2 variables \code{mousegeno}, data frame of 173 observations (genotype) of 5 variables (SNP markers)  and \code{mousepheno}, data frame of 173 observations (normalized gene expression) of 14 variables (genes).\\

\noindent Load the dataset and view the first 3 observations:

<<dataset,results=verbatim>>=
data(mouse,package="geneNetBP")
head(mousegeno,n=3)
head(mousepheno,n=3)
@

\noindent Note that there are 3 possible genotype states MM (homozygous) denoted by 1, H (heterozygous) by 2 and SS (homozygous) by 3. The genotypes are categorical variables and hence all columns in data frame \code{mousegeno} have to be of class factor while the phenotypes are continuous variables with all columns in data frame \code{mousepheno} of class numeric.\\ 


\subsection{toy} 
\noindent The \code{toy} is a simulated eQTL dataset from the network shown below, of 500 observations, 3 genotypes (Q1,Q2,Q3) each having 2 possible states (\code{toygeno}) and 6 phenotypes, X1-X6 (\code{toypheno}).  \\

\begin{figure}[H]
  \centering
  \includegraphics[trim=0 450 300 80,clip,height=7cm,width=7cm]{toy_graph}
  \caption{Toy network example.}
\end{figure}


\subsection{yeast}

\noindent The \code{yeast} dataset is a subset of the widely studied yeast expression dataset comprising of 112 F1 segregants from a cross between BY4716 and RM11-1a strains of \emph{Saccharomyces Cerevisiae} \cite{kruglak2005, kruglak2005nature}. The original dataset consists of expression values reported as log2(sample/ BY reference) for 6216 genes. The data can be accessed in Gene Expression Omnibus (GEO) by accession number (GSE1990). After linkage analysis and filtering based on location and significance of QTL, a final set of 25 genes and their corresponding 12 SNP markers were identified and included in the yeast dataset.\\ 

\noindent Thus the final dataset has 2 variables \code{yeastgeno}, data frame of 112 observations (genotype) of 9 variables (SNP markers)  and \code{yeastpheno}, data frame of 112 observations (normalized gene expression) of 16 variables (genes).\\

\noindent Load the dataset and view the first 3 observations:

<<dataset,results=verbatim,eval=FALSE>>=
data(yeast,package="geneNetBP")
head(yeastgeno,n=3)
head(yeastpheno,n=3)
@

\noindent Note that there are 2 possible genotype states denoted by 1 and 2. The genotypes are categorical variables and hence all columns in data frame \code{yeastgeno} have to be of class factor while the phenotypes are continuous variables with all columns in data frame \code{yeastpheno} of class numeric.\\ 


\end{document}
